File: lintcode/137_clone-graph/clone-graph.cpp
/*
@Copyright:LintCode
@Author:   chenweikang
@Problem:  http://www.lintcode.com/problem/clone-graph
@Language: C++
@Datetime: 16-07-24 21:22
*/

/**
 * Definition for undirected graph.
 * struct UndirectedGraphNode {
 *     int label;
 *     vector<UndirectedGraphNode *> neighbors;
 *     UndirectedGraphNode(int x) : label(x) {};
 * };
 */

class Solution {
public:
    UndirectedGraphNode *cloneGraph(UndirectedGraphNode *node) {
        
        if (!node) return NULL;
        
        UndirectedGraphNode* cp_head = new UndirectedGraphNode(node->label);
        unordered_map<UndirectedGraphNode*, UndirectedGraphNode*> hs;
        hs[node] = cp_head;
        
        queue<UndirectedGraphNode*> q;
        q.push(node);
        
        unordered_set<UndirectedGraphNode*> visited;
        visited.insert(node);
        
        while (!q.empty()) {
            UndirectedGraphNode* cur = q.front(); q.pop();
            UndirectedGraphNode* cp_cur = hs[cur];
            for (auto &one: cur->neighbors) {
                auto it = hs.find(one);
                if (it == hs.end()) 
                    it = hs.insert(make_pair(one, new UndirectedGraphNode(one->label))).first;
                cp_cur->neighbors.push_back(it->second);
                
                if (visited.insert(one).second) q.push(one);
            }
        }
        
        return cp_head;
    }
};

File: lintcode/137_clone-graph/clone-graph.py
# coding:utf-8
'''
@Copyright:LintCode
@Author:   chenweikang
@Problem:  http://www.lintcode.com/problem/clone-graph
@Language: Python
@Datetime: 16-07-24 21:32
'''

# Definition for a undirected graph node
# class UndirectedGraphNode:
#     def __init__(self, x):
#         self.label = x
#         self.neighbors = []
class Solution:
    # @param node, a undirected graph node
    # @return a undirected graph node
    def __init__(self):
        self.dict = {}
        
    def cloneGraph(self, node):
        if not node: return None
        
        cp_head = UndirectedGraphNode(node.label)
        stk = [node]
        hs = {node:cp_head}
        
        while stk:
            cur = stk.pop() 
            cp_cur = hs[cur]
            for one in cur.neighbors:
                if one not in hs:
                    stk.append(one)
                    hs[one] = UndirectedGraphNode(one.label)
                cp_cur.neighbors.append(hs[one])
        return cp_head
        